### ISSUE TITLE
Custom error log format

### ISSUE THREAD

**Issue by bedis** [haproxy/haproxy#826]
## What should haproxy do differently? Which functionality do you think we should add?

HAProxy allows us to format the "traffic" log line the way we want through the `log-format` directive. Many people use it to format the log line in JSON and inject it into tools that can parse it natively.

## What are you trying to do?

some log lines related to traffic processing are not formated using this method, which makes sense. In example, the "SSL handshake error" log line can't be formatted in JSON and so is ignored by the log processing server. That said, the information provided by this log line can be valuable. It would be great to be able to format these log lines the way we want so they can be processed properly.

**Comment by wlallemand** [comment]
This is related to issue #693, I don't know if there is any other case where we can't customize the logs.

**Comment by Zarne** [comment]
in my case, I use logstash to send my log to elasticsearch with the follow config file:
```
filter {
  grok {
    match => [
      "message",
      "%{TIMESTAMP_ISO8601:syslog_timestamp} %{SYSLOGHOST:source_host} (?:%{PROG:program}(?:\[%{POSINT:pid}\])?: )?%{GREEDYDATA:message}"
    ]
    overwrite => [ "message" ]
  }
  syslog_pri { }
  json {
    source => "message"
  }
}
``` 
and my `log-format` is:
```
  log-format  {\"type\":\"haproxy\",\"virtual\":\"%f\",\"client_address\":\"%ci\",\"vip\":\"%fi\",\"client_ip\":\"%ci\",\"client_port\":\"%cp\",\"lb_server\":\"%s\",\"vhost\":"%{+Q}[capture.req.hdr(2)]",\"uri\":\"%HP\",\"querystring\":\"%HQ\",\"stage\":\"prod\",\"mapping\":\"%b\",\"request\":\"%r\",\"verb\":\"%HM\",\"response\":\"%ST\",\"content_type\":"%{+Q}[capture.req.hdr(1)]",\"content_length\":\"%B\",\"referer\":"%{+Q}[capture.req.hdr(3)]",\"x_forwarded_for\":"%{+Q}[capture.req.hdr(0)]",\"resp_time\":\"%Tr\",\"timer\":\"%Th/%Ti/%TR/%Tw/%Tc/%Td\",\"user_agent\":"%{+Q}[capture.req.hdr(4)]",\"x_ftven_uuid\":"%{+Q}[capture.req.hdr(5)]",\"timestamp\":\"%Ts\"}
```
when haproxy send a error log, i have a parse error on logstash because the error log is not in json type.

**Comment by wlallemand** [comment]
Please take a look at ticket #693 about this, a new option `option log-error-via-logformat` was integrated in 2.5-dev which should meet your needs.

**Comment by wtarreau** [comment]
Note for future dev on this, we'll also need to determine how the field names are to be provided, either they're hard-coded in haproxy for each metric, or we provide a default set that may be overridden.

**Comment by wtarreau** [comment]
Related issue: #401

**Comment by wlallemand** [comment]
I'm closing this, since it was implemented in 2.5.
